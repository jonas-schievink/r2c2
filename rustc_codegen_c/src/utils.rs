use std::io::{self, Write};
use std::fmt::{self, Display};
use std::str;
use std::ops::Deref;
use std::ops::DerefMut;
use rustc::ty::TyCtxt;

// Import the host/target triple, set by the build script
include!(concat!(env!("OUT_DIR"), "/triples.rs"));

/// Extension methods for `Result<T, E>`.
pub trait ResultExt<T, E> {
    /// Turns the error into a `String`.
    fn err_to_string(self) -> Result<T, String>
    where
        E: ToString;
}

impl<T, E: Display> ResultExt<T, E> for Result<T, E> {
    fn err_to_string(self) -> Result<T, String> {
        self.map_err(|e| e.to_string())
    }
}

/// Obtains the version information of the system's C compiler.
///
/// TODO: Upstream this to the `cc` crate
pub fn cc_version() -> Result<String, String> {
    // FIXME: cc::Error doesn't impl Display and Error
    let mut compiler = cc::Build::new()
        .try_get_compiler()
        .map_err(|e| format!("{:?}", e))?
        .to_command();
    let output = compiler.arg("--version").output().err_to_string()?;

    let mut out = String::from_utf8(output.stdout).err_to_string()?;
    out.push_str(str::from_utf8(&output.stderr).err_to_string()?);

    if output.status.success() {
        Ok(out)
    } else {
        Err(out)
    }
}


/// Trait for writing UTF-8 data to a sink.
pub trait WriteStr {
    fn write_fmt(&mut self, args: fmt::Arguments) -> io::Result<()>;

    fn write_str(&mut self, s: &str) -> io::Result<()> {
        write!(self, "{}", s)
    }
}

impl<W: Write> WriteStr for W {
    fn write_fmt(&mut self, args: fmt::Arguments) -> io::Result<()> {
        <W as Write>::write_fmt(self, args)
    }
}

pub struct StringWriter(pub String);

impl WriteStr for StringWriter {
    fn write_fmt(&mut self, args: fmt::Arguments) -> io::Result<()> {
        Ok(<String as fmt::Write>::write_fmt(&mut self.0, args).expect("writing to string failed"))
    }
}

impl Deref for StringWriter {
    type Target = String;

    fn deref(&self) -> &String {
        &self.0
    }
}

impl DerefMut for StringWriter {
    fn deref_mut(&mut self) -> &mut String {
        &mut self.0
    }
}

impl AsRef<str> for StringWriter {
    fn as_ref(&self) -> &str {
        &self.0
    }
}
