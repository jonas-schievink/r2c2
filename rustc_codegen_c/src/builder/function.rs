//! Function builder.

use super::types::{FnSig, TypeRef};
use super::Symbol;
use utils::WriteStr;

use hashbrown::HashSet;
use toolshed::Arena;

use std::io;

/// A local variable.
///
/// These can be created in 2 ways: By declaring a function taking arguments
/// (which creates `Variable`s corresponding to the arguments), or by calling
/// `declare_variable` on an existing `FunctionBuilder`, which declares a new
/// local variable inside the function.
#[derive(Debug, Copy, Clone)]
pub struct Variable<'a> {
    name: &'a str,
    ty: TypeRef<'a>,
}

/// Builder for function bodies.
///
/// Created by `TranslationUnitBuilder::define_function`.
pub struct FunctionBuilder<'a, W: WriteStr> {
    /// Output statements are written to this writer.
    writer: &'a mut W,
    arena: &'a Arena,
    declared_locals: HashSet<String>,
    finished: bool,

    /// Function arguments, available as declared `Variable`s.
    pub args: &'a [Variable<'a>],
}

impl<'a, W: WriteStr> FunctionBuilder<'a, W> {
    /// Creates a function builder, writing the function's header to `writer`.
    ///
    /// **Note**: You *must* call `finish` on the builder to close the function
    /// being built, or this will panic on drop.
    pub fn create(
        writer: &'a mut W,
        arena: &'a Arena,
        name: Symbol<'_>,
        proto: FnSig<'a>,
    ) -> io::Result<Self> {
        proto.declare(name.mangled, writer)?;
        writeln!(writer)?;
        writeln!(writer, "{{")?;
        Ok(Self {
            writer,
            arena,
            declared_locals: HashSet::new(),
            finished: false,
            args: arena.alloc_slice(
                &proto
                    .args
                    .iter()
                    .enumerate()
                    .map(|(i, &ty)| Variable {
                        name: arena.alloc_str(&format!("_{}", i)),
                        ty,
                    })
                    .collect::<Vec<_>>(),
            ),
        })
    }

    /// Closes the in-progress function definition, consuming `self`.
    pub fn finish(mut self) -> io::Result<()> {
        writeln!(self.writer, "}}")?;
        writeln!(self.writer)?;
        self.finished = true;
        Ok(())
    }

    fn indent(&mut self) -> io::Result<()> {
        write!(self.writer, "    ")
    }

    /// Declares a new local variable at the current position.
    pub fn declare_variable<'b>(
        &mut self,
        name: impl AsRef<str>,
        ty: TypeRef<'a>,
        comment: impl Into<Option<&'b str>>,
    ) -> io::Result<Variable<'a>> {
        let name = name.as_ref();
        self.indent()?;
        ty.declare_variable(name, self.writer)?;
        write!(self.writer, ";")?;
        if let Some(comment) = comment.into() {
            write!(self.writer, "  /* {} */", comment)?;
        }
        writeln!(self.writer)?;

        Ok(Variable {
            name: self.arena.alloc_str(name),
            ty,
        })
    }
}

impl<'a, W: WriteStr> Drop for FunctionBuilder<'a, W> {
    fn drop(&mut self) {
        if !self.finished {
            panic!("FunctionBuilder dropped without calling `finish`");
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use builder::test::compile_test;

    #[test]
    fn declare_locals() {
        compile_test("declare_locals", |tu| {
            let a = Arena::new();
            let sig = tu.fn_sig(None, &[]);
            let double = tu.double();
            let fnptr = tu.fn_ptr(sig);
            let mut f = tu.define_function(&a, Symbol::test("declare_locals"), sig)?;
            f.declare_variable("dbl", double, None)?;
            f.declare_variable("f", fnptr, Some("i'm a function pointer with a comment"))?;
            f.finish()?;
            Ok(())
        });
    }
}
