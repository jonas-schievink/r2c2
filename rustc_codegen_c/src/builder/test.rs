//! Unit test utilities.

#![cfg(test)]

use super::function::FunctionBuilder;
use super::{FnSig, Symbol, TranslationUnitBuilder, TypeRef};
use utils::{StringWriter, WriteStr};

use cc::Build;
use std::error::Error;
use std::io::Write;
use syntax_pos::edition::DEFAULT_EDITION;
use syntax_pos::{Globals, GLOBALS};
use tempfile::{self, TempDir};
use toolshed::Arena;

/// Result to return from unit tests (instead of panicking).
pub type TestResult = Result<(), Box<Error>>;

/// Calls the test function `f` with a `TranslationUnitBuilder` and tests
/// that the produced output can be compiled by the system's C compiler.
///
/// The output is also snapshot-tested using `insta`.
pub fn compile_test<F>(name: &str, f: F)
where
    F: FnOnce(&mut TranslationUnitBuilder<'_, StringWriter>) -> TestResult,
{
    GLOBALS.set(&Globals::new(DEFAULT_EDITION), || {
        let arena = Arena::new();
        let mut writer = StringWriter(String::new());
        let mut builder = TranslationUnitBuilder::create(&arena, writer).unwrap();

        f(&mut builder).unwrap();

        let dir = TempDir::new().unwrap();

        let mut f = tempfile::Builder::new().suffix(".c").tempfile().unwrap();
        f.write_all(&builder.writer().0.as_bytes()).unwrap();
        Build::new()
            .file(f.path())
            .cargo_metadata(false)
            .out_dir(&dir)
            .target(::utils::TARGET)
            .host(::utils::HOST)
            .opt_level(0)
            .compile("foo");
        let output = builder.into_writer().0;
        assert_snapshot_matches!(name, output);
    })
}
