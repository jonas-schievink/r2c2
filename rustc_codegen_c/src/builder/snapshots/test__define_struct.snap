---
created: "2019-12-08T16:29:55.795006431Z"
creator: insta@0.7.4
source: rustc_codegen_c/src/builder/test.rs
expression: output
---
/* generated by r2c2 version 0.0.0 */
#include <stdint.h>
#include <stdbool.h>

/*                                                    */
/*                    END PREAMBLE                    */
/*                                                    */

struct my_struct;  /* (test-symbol: no associated Rust name) */
union my_union;  /* (test-symbol: no associated Rust name) */
struct my_struct {  /* (test-symbol: no associated Rust name) */
    uint32_t intfield;
    uint32_t * intptr;
    void * voidptr;
    bool shevik;
};

